src/dataflow/src/render/context.rs

```diff
@@ -187,22 +194,119 @@ where
         I: IntoIterator,
         I::Item: Data,
         L: FnMut(&V) -> I + 'static,
+        K: FnMut(&[ScalarExpr]) -> Option<V>,
     {
+        use differential_dataflow::{
+            trace::{BatchReader, Cursor},
+            AsCollection,
+        };
+        use timely::dataflow::channels::pact::Pipeline;
+        use timely::dataflow::operators::Operator;
+
         if let Some((oks, err)) = self.collections.get(relation_expr) {
             Some((oks.flat_map(move |v| logic(&v)), err.clone()))
         } else if let Some(local) = self.local.get(relation_expr) {
-            let (oks, errs) = local.values().next().expect("Empty arrangement");
-            Some((
-                oks.flat_map_ref(move |_k, v| logic(v)),
-                errs.as_collection(|k, _v| k.clone()),
-            ))
+            // Determine a key that is constrained to a literal value.
+            let mut constrained_keys = local
+                .keys()
+                .filter_map(|key| key_selector(key).map(|row| (key.clone(), row)))
```

Comment:
We can allow for code reuse b/w the local/trace branches by creating a templated submethod a la `join.rs::differential_join_inner<J, Tr2>`

Reply:
Done!